pkg/image/registryclient/client.go

```diff
@@ -226,6 +249,46 @@ func (c *Context) Repository(ctx context.Context, registry *url.URL, repoName st
 	return NewLimitedRetryRepository(repo, c.Retries, limiter), nil
 }
 
+// RepositoryWithAlternateReference returns a a valid repository and ref, it gathers alternate sources then returns first repo, ref, or error
+func (c *Context) RepositoryWithAlternateReference(ctx context.Context, locator imagereference.DockerImageReference, insecure bool) (distribution.Repository, imagereference.DockerImageReference, error) {
```

Comment:
Since the point of FirstRequest is to be called the first time get is invoked, I don't see how this code knows that this is the "first request", nor do I see the logic I asked for before that shows this in the context of a specific method.

Who calls this?

Reply:
To help frame what I'm asking let me take a stab at implementing the logic inline for the wrappers that I expect, then we can circle back

Reply:
This is called by `oc` (pkg/cli/image/imagesource/options.go, replacing the `regcontext.Repository`) [link to oc PR](https://github.com/openshift/oc/pull/439/files#diff-22e127520b7182edf749daf679f1963186ce0e55e13edf001780b06f4b59d4f7)

Reply:
I asked for oc not to need to do that though - that means that multiple callers have to know about this.

I tried to capture this into a semi working form in https://github.com/openshift/library-go/pull/1021 - see how methods like Get and Open are retried (by calling the alternates, which defers and shares the source with multiple invocations), while calls like Put and Create are only invoked once.  Once you get a chance to look through that code, we should schedule some time and see if we can work through where gaps may be between this PR and my example code.

Reply:
Image-registry will try the mirrored references first, b4 the source ref, which is why I am gathering sources in that order . They don't afaict have a need for FirstRequest, and neither does `oc` at least with the issues that i've been working to resolve. I don't have enough context to implement this - you suggested earlier to add it and for the simple strategy return nil sources & nil error, which is what I have currently done - I'll need to know more about a use case for this in order to implement it further.

Reply:
^^ i did not see your reply when i wrote that, did not refresh the page.. i'm looking at your example now

Reply:
It's taking me a bit of time to refactor this to include clayton's suggestion, but I'm making progress,  and will reach out if/when stuck.

Reply:
ok i've pushed w/ clayton's #1021, ptal